[KBFS-1185] Use per-TLF storage for MDServerDisk

akalin-keybase opened the pull request:
Split up MDServerDisk into per-TLF objects (mdServerTlfStorage).
Make those objects store RootMetadataSigned objects in flat files.
They will eventually store the rest of the data (currently in
levelDBs) in a journal.

Fix up naming in bserver_disk.go and bserver_tlf_journal.go.

Also check for shutdown only in public BServer/MDServer
methods.

Make md.MetadataID() take only a crypto object.

Make MDServer.GetForHandle serialize TLF ID creation.

akalin-keybase commented on .travis.yml:
```diff
@@ -42,7 +42,6 @@ script:
   # Run libkbfs tests with an in-memory bserver and mdserver, and run all other
   # tests with the tempdir bserver and mdserver.
   - cd libkbfs && go test -i && go test -race -c && ./libkbfs.test -test.timeout 2m
-  - env KEYBASE_TEST_BSERVER_ADDR=tempdir ./libkbfs.test -test.timeout 2m
```
Removed this, to match the comments.

akalin-keybase commented:
Okay, this is ready for review.

strib commented on libkbfs/mdserver_disk.go:
```diff
@@ -5,97 +5,66 @@
-var _ mdServerLocal = (*MDServerDisk)(nil)
+// MDServerDisk stores all info on disk.
+type MDServerDisk struct {
+	codec  Codec
```
When there's more than one of these config-originating objects, I'm not a big fan of storing them individually.  We have some tests that swap things into the `Config` object on the fly, and I can imagine a time when that happens in production too, especially for things like crypto and kbpki.  Plus it seems more difficult for the test to construct four separate things, rather than just using a standard test config.  Can we change it back? Or am I missing something about why this is better?

akalin-keybase replied:
Done, although I'd like to keep them for the per-TLF storage, since those only store codec and cryptoPure, which shouldn't change at all.

strib commented on libkbfs/mdserver_disk.go:
```diff
@@ -5,97 +5,66 @@
-	mdserv := &MDServerDisk{config, handleDb, mdDb, branchDb, log,
-		&sync.Mutex{}, locksDb, newMDServerLocalUpdateManager(),
-		&sync.RWMutex{}, new(bool), shutdownFunc}
+	mdserv := &MDServerDisk{config.Codec(), config.Clock(),
```
Slight preference for using field names in the constructor here, to avoid having to do `sync.RWMutex{}`.

akalin-keybase replied:
Done.

strib commented on libkbfs/mdserver_disk.go:
```diff
@@ -120,445 +89,176 @@ func NewMDServerTempDir(config Config) (*MDServerDisk, error) {
-		return nil, MDServerErrorUnauthorized{}
+
+	md.lock.Lock()
+	defer md.lock.Unlock()
```
Would it be safe to unlock immediately after the nil check below, so we don't hold the lock while blocking on leveldb access?

strib replied:
Nevermind, that's a bad idea.

strib commented on libkbfs/mdserver_tlf_storage.go:
```diff
@@ -0,0 +1,596 @@
+
+	// Store the truncate lock holder just in memory, so it gets
+	// wiped after a restart.
+	truncateLockHolder *keybase1.KID
```
Just a heads up -- we'll want to plumb truncate calls through to the real MD server (or redirect them at a higher level) even when we're in async write mode.    I guess we'll still want support for it here for tests though.  No suggestions on things to change now, just mentioning it for the future.

akalin-keybase replied:
Yeah, I actually moved this back to be handled by `MDServerLocal`, since I don't think the journal needs to know about those directly.

strib commented on libkbfs/mdserver_tlf_storage.go:
```diff
@@ -0,0 +1,596 @@
+	}
+	buf := &bytes.Buffer{}
+
+	// this order is signifcant for range fetches.
```
Not your fault, but "signifcant" -> "significant" while you're here?

akalin-keybase replied:
Done.

strib commented:
Looks good to me, thanks!

akalin-keybase commented:
Addressed all comments, but realized from working on the follow-up PR that I put too much stuff in `mdServerTlfStorage`, so I moved some stuff back to `MDServerDisk`. In particular, the truncate locks (which do the same as `MDServerMemory`) and the `branchDB` table, so that `mdServerTlfStorage` just has to know about branch IDs. I think I made enough changes to warrant another look, so PTAL!

akalin-keybase commented:
Rebased this on top of https://github.com/keybase/kbfs/pull/102 , although I think that's small enough that you can still re-review this without waiting for that one to go in.

strib commented on libkbfs/mdserver_local_shared.go:
```diff
@@ -61,6 +63,51 @@ func isWriterOrValidRekey(ctx context.Context, codec Codec, kbpki KBPKI,
 
+type mdServerLocalTruncateLockManager struct {
+	// TLF ID -> device KID.
+	locksDb map[TlfID]keybase1.KID
```
No mutex for the map?  If you're relying on the higher level mutex, a comment would probably be good.

akalin-keybase replied:
Done, added comments.

strib commented:
Looks good to me, thanks!
